File: src/main/java/ru/romanow/jpa/mapper/PersonMapper.java
package ru.romanow.jpa.mapper;

import org.mapstruct.*;
import ru.romanow.jpa.domain.Person;
import ru.romanow.jpa.model.PersonModifyRequest;
import ru.romanow.jpa.model.PersonResponse;

import static org.mapstruct.NullValuePropertyMappingStrategy.IGNORE;

@Mapper(config = MapperConfiguration.class,
        uses = { AddressMapper.class, AuthorityMapper.class, RoleMapper.class, ReferenceMapper.class })
public interface PersonMapper {

    PersonResponse toModel(Person person);

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "addressId", ignore = true)
    Person toEntity(PersonModifyRequest response);

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "addressId", ignore = true)
    @Mapping(target = "roles", ignore = true)
    @Mapping(target = "address", ignore = true)
    @Mapping(target = "authorities", ignore = true)
    @BeanMapping(nullValuePropertyMappingStrategy = IGNORE)
    void update(PersonModifyRequest request, @MappingTarget Person person);

    @InheritConfiguration(name = "toEntity")
    @Mapping(target = "address", qualifiedBy = FullUpdate.class)
    void fullUpdate(PersonModifyRequest request, @MappingTarget Person person);
}

File: src/main/java/ru/romanow/jpa/mapper/PersonMapper.kt
package ru.romanow.jpa.mapper

import org.mapstruct.*
import org.mapstruct.NullValuePropertyMappingStrategy.IGNORE
import ru.romanow.jpa.domain.Person
import ru.romanow.jpa.mapper.config.MapperConfiguration
import ru.romanow.jpa.mapper.utils.ReferenceMapper
import ru.romanow.jpa.model.PersonModifyRequest
import ru.romanow.jpa.model.PersonResponse

@Mapper(
    config = MapperConfiguration::class,
    collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED,
    uses = [AddressMapper::class, AuthorityMapper::class, RoleMapper::class, ReferenceMapper::class]
)
interface PersonMapper {
    fun toModel(person: Person): PersonResponse

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "addressId", ignore = true)
    fun toEntity(response: PersonModifyRequest): Person

    @InheritConfiguration(name = "toEntity")
    @BeanMapping(nullValuePropertyMappingStrategy = IGNORE)
    fun update(request: PersonModifyRequest, @MappingTarget person: Person)
}
